File: frontend/lib/reports.ts
// Reports client — daily summary + week-over-week trends.

import { api } from "./api";

export interface DailyRow {
  date: string;
  joined: number;
  seated: number;
  walked_away: number;
  walkaway_rate: number | null;
  premium_sold: number;
  premium_revenue: number;
  median_wait_mins: number | null;
  median_wait_regular: number | null;
  median_wait_premium: number | null;
  premium_wait_saving_mins: number | null;
  walkaway_spike: boolean;
}

export interface WeekComparison {
  seated: number;
  walked_away: number;
  premium_sold: number;
  premium_revenue: number;
  median_wait_mins: number | null;
}

export interface DailyReport {
  days: number;
  rows: DailyRow[];
  this_week: WeekComparison;
  prior_week: WeekComparison;
}

export const reportsApi = {
  daily: (token: string, days = 28) =>
    api<DailyReport>(`/api/reports/daily?days=${days}`, { token }),
};

export interface StatementLine {
  date: string;
  time: string;
  ticket_no: number | null;
  party_size: number;
  gross_amount: number;
  restaurant_share: number;
  ifasto_fee: number;
}

export interface MonthlyStatement {
  month: string;
  venue_name: string;
  lines: StatementLine[];
  passes_sold: number;
  gross_total: number;
  restaurant_total: number;
  ifasto_total: number;
}

export const statementApi = {
  get: (token: string, month?: string) =>
    api<MonthlyStatement>(
      `/api/reports/statement${month ? `?month=${month}` : ""}`,
      { token }
    ),
};

File: backend/app/schemas/reports.py
"""Schemas for the daily/trends report — the operator's renewal artifact."""

from __future__ import annotations

from pydantic import BaseModel


class DailyRow(BaseModel):
    date: str  # JST date, ISO
    joined: int
    seated: int
    walked_away: int
    walkaway_rate: float | None  # walked / (seated + walked), None if no outcomes
    premium_sold: int
    premium_revenue: int  # yen, from Transaction rows
    median_wait_mins: float | None  # seated parties, all types
    median_wait_regular: float | None
    median_wait_premium: float | None
    premium_wait_saving_mins: float | None  # regular median - premium median
    walkaway_spike: bool  # rate > 1.5x window mean (min 3 walk-aways)


class WeekComparison(BaseModel):
    seated: int
    walked_away: int
    premium_sold: int
    premium_revenue: int
    median_wait_mins: float | None


class DailyReport(BaseModel):
    days: int
    rows: list[DailyRow]
    this_week: WeekComparison
    prior_week: WeekComparison


class StatementLine(BaseModel):
    date: str  # JST date of sale, ISO
    time: str  # JST HH:MM
    ticket_no: int | None
    party_size: int
    gross_amount: int  # yen
    restaurant_share: int  # 70%
    ifasto_fee: int  # 30%


class MonthlyStatement(BaseModel):
    month: str  # YYYY-MM (JST)
    venue_name: str
    lines: list[StatementLine]
    passes_sold: int
    gross_total: int
    restaurant_total: int
    ifasto_total: int
